opm/simulators/linalg/bda/BdaBridge.cpp

```diff
@@ -164,27 +165,27 @@ int checkZeroDiagonal(BridgeMatrix& mat) {
 // iterate sparsity pattern from Matrix and put colIndices and rowPointers in arrays
 // sparsity pattern should stay the same
 // this could be removed if Dune::BCRSMatrix features an API call that returns colIndices and rowPointers
-template <class BridgeMatrix>
-void getSparsityPattern(BridgeMatrix& mat, std::vector<int> &h_rows, std::vector<int> &h_cols) {
+template <class BridgeMatrix, class BridgeVector, int block_size>
+void BdaBridge<BridgeMatrix, BridgeVector, block_size>::getSparsityPattern(const BridgeMatrix& mat, int *h_rows, int *h_cols) {
     int sum_nnzs = 0;
+    int nnz_idx = 0;
 
     // convert colIndices and rowPointers
-    if (h_rows.empty()) {
-        h_rows.emplace_back(0);
-            for (typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r) {
-                int size_row = 0;
-                for (auto c = r->begin(); c != r->end(); ++c) {
-                    h_cols.emplace_back(c.index());
-                    size_row++;
-                }
-                sum_nnzs += size_row;
-                h_rows.emplace_back(sum_nnzs);
-            }
-
-        // h_rows and h_cols could be changed to 'unsigned int', but cusparse expects 'int'
-        if (static_cast<unsigned int>(h_rows[mat.N()]) != mat.nonzeroes()) {
-            OPM_THROW(std::logic_error, "Error size of rows do not sum to number of nonzeroes in BdaBridge::getSparsityPattern()");
+    h_rows[0] = 0;
+    for (typename BridgeMatrix::const_iterator r = mat.begin(); r != mat.end(); ++r) {
+        int size_row = 0;
+        for (auto c = r->begin(); c != r->end(); ++c) {
+            h_cols[nnz_idx] = c.index();
+            size_row++;
+            nnz_idx++;
         }
+        sum_nnzs += size_row;
+        h_rows[r.index() + 1] = sum_nnzs;
+    }
```

Comment:
Probably will not find the time for a full review this week, maybe we will/shoudl discuss on Monday at our meeting.

Just a note here (in case I might forget):
I don't see why saw a need to change this. Please never resort to C if you can use C++ (especially if you are actually changing C++ containers). 
@bska is completely right with his concern. You are just reserving and assuming that for an empty vector with size 0 but reserved space `data()` will return the pointer to he reserved space. Unfortunately, the standard allows returning a nullpointer.

In addition just reserving space only worked before because you used `emplace_back` here as it will adapt the size of the container. With your change to C that is not the case anymore and the vectors will remain at size 0 which will be unexpected to fellow developer (including you in the future).